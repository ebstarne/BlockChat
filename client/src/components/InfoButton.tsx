import React from 'react';
import Dialog from '@material-ui/core/Dialog';
import { useLocation } from 'react-router-dom';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import InfoIcon from '@material-ui/icons/Info';
import IconButton from '@material-ui/core/IconButton';

const InfoButton = (props: any) => {
  const [openInfo, setOpenInfo] = React.useState(false);
  let infoTitle = '';
  let infoText = '';

  const handleInfoOpen = () => {
    setOpenInfo(true);
  };

  const handleInfoClose = () => {
    setOpenInfo(false);
  };
  const route = useLocation().pathname;
  switch (route) {
    case '/':
      infoTitle = 'Welcome!';
      infoText =
        'This is BlockChat! This is a chatting application built to show off how the blockchain works and how it can be used in business operations. While a chatting application may not be the best real world application of the blockchain, it can be used to show off some functionality of it.\n Created by NCSU CSC Senior Design Team 1, Fall 2020:\nDominick Bagnoli, Seth Hollandsworth, Thomas Landsberg, and Eric Starner.';
      break;
    case '/login':
      infoTitle = 'Login';
      infoText =
        'This is the login screen showing you have entered your organization. Notice the entire website theme changed! Here you will need to log in. If you do not have an account for this organization, you can choose to make one with a unique username and password. After you create an account you will be automatically logged in and taken to the Message Center.';
      break;
    case '/message':
      infoTitle = 'Message Center';
      infoText =
        'This is the Message Center, the meat of this application. There is a lot to explain here in terms of the blockchain. The left panel shows the channels your organization belongs to. Organizations can have channels between one or more other organizations on the same blockchain!\nWhen you send a message, the message goes to each organization’s local database first, then is distributed to the blockchain where the transaction is verified by all other organizations in the channel. This means people in your organization will receive your message before others in a different organizations.\nAfter your message appears in your message history, you can click it to see where your message has been and when it took these steps.\nOne of the biggest pieces of information here is that organizations will each have their own database with message history, but the blockchain will hold the “master list” of transactions. Transactions can never be edited or deleted from the blockchain. Another thing to note is that even though organizations will have their own databases, they will still be able to share information and transactions across their separate databases.';
      break;
  }

  return (
    <>
      <IconButton
        aria-label='additional info for screen'
        aria-haspopup='true'
        onClick={handleInfoOpen}
        color='inherit'>
        <InfoIcon />
      </IconButton>
      <Dialog
        aria-labelledby='transition-modal-title'
        aria-describedby='transition-modal-description'
        open={openInfo}
        onClose={handleInfoClose}
        closeAfterTransition
        BackdropProps={{
          timeout: 500,
        }}>
        <DialogTitle>{infoTitle}</DialogTitle>
        <DialogContent>
          <DialogContentText style={{ color: 'black' }}>
            {infoText.split('\n').map((str, index) => (
              <p key={index}>{str}</p>
            ))}
          </DialogContentText>
        </DialogContent>
      </Dialog>
    </>
  );
};
export default InfoButton;
