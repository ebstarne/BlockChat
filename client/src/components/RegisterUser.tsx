import React, { useState } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import Grid from '@material-ui/core/Grid';
import { useSnackbar } from 'notistack';

// Redux
import { connect } from 'react-redux';
import { register } from '../redux/ducks/actions/auth';

const RegisterUser = (props: any) => {
  const { enqueueSnackbar } = useSnackbar();
  const [email, setEmail] = useState('');
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const handleSubmit = async (event: any) => {
    event.preventDefault();
    if (
      (await props.register({
        username,
        email,
        password,
        organization: props.organization,
      })) === 'failure'
    ) {
      enqueueSnackbar('Register Failed', { variant: 'warning' });
    }
  };

  const [open, setOpen] = React.useState(false);

  const handleClickOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  return (
    <div>
      <Button
        variant='contained'
        color='primary'
        disableElevation
        onClick={handleClickOpen}>
        Create New Account
      </Button>
      <Dialog
        open={open}
        onClose={handleClose}
        fullWidth={true}
        maxWidth='xs'
        aria-labelledby='form-dialog-title'>
        <DialogTitle id='form-dialog-title'>Sign Up</DialogTitle>
        <DialogContent>
          <DialogContentText>It's that easy</DialogContentText>
          <form onSubmit={handleSubmit} autoComplete='off'>
            <Grid
              container
              direction='column'
              justify='space-evenly'
              alignItems='center'
              spacing={1}>
              <Grid item>
                <TextField
                  variant='filled'
                  type='text'
                  name='email'
                  label='Email'
                  onChange={(e) => {
                    setEmail(e.target.value);
                  }}
                  value={email}
                />
              </Grid>
              <Grid item>
                <TextField
                  variant='filled'
                  type='username'
                  name='username'
                  label='Username'
                  onChange={(e) => {
                    setUsername(e.target.value);
                  }}
                  value={username}
                />
              </Grid>

              <Grid item>
                <TextField
                  variant='filled'
                  type='password'
                  name='password'
                  label='Password'
                  onChange={(e) => {
                    setPassword(e.target.value);
                  }}
                  value={password}
                />
              </Grid>
            </Grid>
            <DialogActions>
              <Button onClick={handleClose} color='primary'>
                Cancel
              </Button>
              <Button
                onClick={handleClose}
                color='primary'
                type='submit'
                disabled={username === '' || password === ''}>
                Sign Up
              </Button>
            </DialogActions>
          </form>
        </DialogContent>
      </Dialog>
    </div>
  );
};

const mapStateToProps = (state: any) => ({
  isAuthenticated: state.auth.isAuthenticated,
  organization: state.theme.org,
});

export default connect(mapStateToProps, { register })(RegisterUser);
