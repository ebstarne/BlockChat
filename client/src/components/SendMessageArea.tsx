import React, { useEffect, useState, KeyboardEvent } from 'react';
import Drawer from '@material-ui/core/Drawer';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import SendRoundedIcon from '@material-ui/icons/SendRounded';
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';
// normal api calls
import { sendMessage } from '../utils/message';
// models
import Message from '../models/Message';
import User from '../models/User';
// Redux
import { connect } from 'react-redux';

const drawerWidth = 240;
const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    messageSendWindow: {
      left: drawerWidth,
    },
    messageArea: {
      width: '100%',
      height: '15vh',
    },
  })
);
const SendMessageArea: React.FC = (props: any) => {
  const classes = useStyles();
  const [messageValue, setMessageValue] = useState('');
  const [messageSent, setMessageSent] = useState(false);
  // get the data from the current user in the store
  const currentUser: User = props.userInfo;
  // cleanup the text box after a message has been sent
  // was having an issue with the "enter" to send a message staying in the box after the message was sent
  // this is for that scenario
  useEffect(() => {
    setMessageValue('');
  }, [messageSent]);

  // function to add a new message to the channel
  const submitMessage = () => {
    // don't submit the message if the contents are empty
    if (!messageValue || !messageValue.trim()) {
      return;
    }

    const newMessage: Message = {
      user: currentUser,
      username: currentUser.username,
      text: messageValue,
      date: new Date(),
      channel: props.currentChannel.channel,
      organization: currentUser.organization.name,
    };

    // send message to the database
    const response = sendMessage(newMessage, props.auth);

    // say that we have sent a message so the text box can be cleaned up
    setMessageSent(!messageSent);
  };

  // function to determine if enter is being pressed in the message box
  const enterPress = (event: KeyboardEvent) => {
    return event.key === 'Enter' && !event.shiftKey;
  };

  return (
    <>
      <Drawer
        anchor={'bottom'}
        classes={{
          paper: classes.messageSendWindow,
        }}
        variant='permanent'
        open>
        <form
          onSubmit={(e) => {
            e.preventDefault();
            submitMessage();
          }}>
          <Grid container direction='row' className={classes.messageArea}>
            <Grid item xs={11}>
              <TextField
                id='standard-full-width'
                style={{ margin: 8, verticalAlign: 'bottom' }}
                placeholder='Message the Channel'
                fullWidth
                multiline
                rowsMax={4}
                rows={1}
                margin='normal'
                value={messageValue}
                onKeyPress={(e) => {
                  if (enterPress(e)) {
                    submitMessage();
                  }
                }}
                onChange={(e) => {
                  setMessageValue(e.target.value);
                }}
                InputLabelProps={{
                  shrink: true,
                }}
              />
            </Grid>
            <Grid item xs={1}>
              <Button type='submit'>
                <SendRoundedIcon />
              </Button>
            </Grid>
          </Grid>
        </form>
      </Drawer>
    </>
  );
};
const mapStateToProps = (state: any) => ({
  currentChannel: state.channel,
  isAuthenticated: state.auth.isAuthenticated,
  auth: state.auth.token,
  userInfo: state.auth.user,
});

export default connect(mapStateToProps)(SendMessageArea);
