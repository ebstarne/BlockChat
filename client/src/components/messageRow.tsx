import React from 'react';
import Avatar from '@material-ui/core/Avatar';
import ListItemAvatar from '@material-ui/core/ListItemAvatar';
import Popper from '@material-ui/core/Popper';
import Fade from '@material-ui/core/Fade';
import Paper from '@material-ui/core/Paper';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import PopupState, { bindToggle, bindPopper } from 'material-ui-popup-state';
import Message from '../models/Message';
/**
 * helper function to render the message inside the scrollable window
 * @param props
 */
const messageRow = (data: Message[], index: number) => {
  const messageToRender = data[index];
  const event = {
    message: 'testing this message for adding to blockchain',
    time: new Date(),
  };
  const events = [event, event, event, event];
  return (
    <PopupState variant='popper' popupId='popup-popper'>
      {(popupState) => (
        <div>
          <ListItem button key={index} {...bindToggle(popupState)}>
            <ListItemAvatar>
              <Avatar
                alt={'test'}
                src={require(`../assets/avatars/${messageToRender.organization.replace(
                  /\s/g,
                  ''
                )}Avatar.png`)}
              />
            </ListItemAvatar>
            <ListItemText
              primary={`${messageToRender.username}: ${messageToRender.text}`}
              primaryTypographyProps={{
                style: { wordWrap: 'break-word' },
              }}
              secondary={`${new Date(messageToRender.date).toLocaleString()}`}
            />
          </ListItem>

          <Popper
            {...bindPopper(popupState)}
            transition
            style={{ zIndex: 10000 }}>
            {({ TransitionProps }) => (
              <Fade {...TransitionProps} timeout={350}>
                <Paper style={{ borderSpacing: 3 }}>
                  <List>
                    {events.map((event, index) => {
                      return (
                        <>
                          <ListItem key={index}>
                            <ListItemText
                              primary={`${event.message}`}
                              secondary={`${event.time}`}
                            />
                          </ListItem>
                          {index + 1 !== events.length && <hr />}
                        </>
                      );
                    })}
                  </List>
                </Paper>
              </Fade>
            )}
          </Popper>
        </div>
      )}
    </PopupState>
  );
};

export default messageRow;
