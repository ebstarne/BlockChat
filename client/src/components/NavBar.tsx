import React from 'react';
import { Redirect } from 'react-router-dom';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import InfoButton from './InfoButton';
import state from '../assets/state.png';

import { connect } from 'react-redux';
import { logout } from '../redux/ducks/actions/auth';

const NavBar = (props: any) => {
  const logoutUser = () => {
    // logs out logically
    props.logout();
    // brings you back to the home screen
    // Message Center already has this logic but if you want to add other screens you can log out
    // this'll be needed
    return <Redirect to='/' />;
  };

  return (
    <>
      <AppBar position='relative' style={{ boxShadow: 'none' }}>
        <Toolbar>
          <Grid
            container
            direction='row'
            justify='space-between'
            alignItems='center'>
            <Grid item>
              <img
                src={state}
                alt='State'
                width='143'
                height='64'
                style={{ alignContent: 'left' }}
              />
            </Grid>
            <Grid item>
              <div>
                <InfoButton />
                {props.auth && (
                  <Button
                    onClick={logoutUser}
                    variant='contained'
                    style={{ zIndex: 10000 }}>
                    Logout
                  </Button>
                )}
              </div>
            </Grid>
          </Grid>
        </Toolbar>
      </AppBar>
    </>
  );
};

const mapStateToProps = (state: any) => ({
  auth: state.auth.token,
});

export default connect(mapStateToProps, { logout })(NavBar);
