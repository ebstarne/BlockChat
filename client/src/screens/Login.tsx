import React, { useState } from 'react';
import nc from '../assets/nc.png';
import { makeStyles } from '@material-ui/core/styles';
import TextField from '@material-ui/core/TextField';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import RegisterUser from '../components/RegisterUser';
import NavBar from '../components/NavBar';
import { Typography } from '@material-ui/core';
import { useSnackbar } from 'notistack';

// Redux
import { Redirect } from 'react-router-dom';
import { connect } from 'react-redux';
import { login } from '../redux/ducks/actions/auth';

const useStyles = makeStyles((theme) => ({
  root: {
    height: '90vh',
  },
  paper: {
    margin: theme.spacing(8, 4),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'left',
    position: 'absolute',
    right: '1%',
    top: '25%',
  },
  image: {
    backgroundImage: `url(${nc})`,
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'cover',
    backgroundPosition: 'center',
  },
  formControl: {
    minWidth: 195,
  },
}));

const Login = (props: any) => {
  const { enqueueSnackbar } = useSnackbar();
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const handleSubmit = async () => {
    if ((await props.login(username, password)) === 'failure') {
      enqueueSnackbar('Login Failed', { variant: 'warning' });
    }
  };
  const classes = useStyles();

  // Redirect if logged in
  if (props.isAuthenticated) {
    return <Redirect to='/message' />;
  }

  return (
    <React.Fragment>
      <NavBar />
      <Grid container className={classes.root}>
        <Grid
          item
          className={classes.image}
          xs={false}
          sm={7}
          md={8}
          lg={9}></Grid>

        <Grid
          className={classes.paper}
          container
          item
          direction='column'
          justify='space-evenly'
          alignItems='flex-start'
          alignContent='flex-end'
          spacing={1}
          xs={12}
          sm={5}
          md={4}
          lg={3}>
          <Grid item>
            <Typography variant='h5'>{props.organization}</Typography>
          </Grid>
          <Grid item>
            <TextField
              variant='filled'
              type='username'
              name='username'
              label='Username'
              onChange={(e) => {
                setUsername(e.target.value);
              }}
              value={username}
            />
          </Grid>

          <Grid item>
            <TextField
              variant='filled'
              type='password'
              name='password'
              label='Password'
              onChange={(e) => {
                setPassword(e.target.value);
              }}
              value={password}
            />
          </Grid>

          <Grid item>
            <Button
              variant='contained'
              color='primary'
              type='submit'
              disableElevation
              disabled={username === '' || password === ''}
              onClick={handleSubmit}>
              Log In
            </Button>
          </Grid>

          <Grid item>
            <RegisterUser />
          </Grid>
        </Grid>
      </Grid>
    </React.Fragment>
  );
};

const mapStateToProps = (state: any) => ({
  isAuthenticated: state.auth.isAuthenticated,
  organization: state.theme.org,
});

export default connect(mapStateToProps, { login })(Login);
