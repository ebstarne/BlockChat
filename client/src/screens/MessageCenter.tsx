import React, { useEffect, useState, useRef } from 'react';
// material ui
import Badge from '@material-ui/core/Badge';
import Divider from '@material-ui/core/Divider';
import Drawer from '@material-ui/core/Drawer';
import Hidden from '@material-ui/core/Hidden';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import StopRoundedIcon from '@material-ui/icons/StopRounded';
import ListItemText from '@material-ui/core/ListItemText';
// our stuff
import messageRow from '../components/messageRow';
import SendMessageArea from '../components/SendMessageArea';
import NavBar from '../components/NavBar';

import { Redirect } from 'react-router-dom';
import { makeStyles, Theme, createStyles } from '@material-ui/core/styles';
// pictures
import state from '../assets/state.png';
// state-changing api calls
import { changeChannel } from '../redux/ducks/actions/channel';
// normal api calls
import { getMessagesFromChannel } from '../utils/message';
import { getChannels } from '../utils/channel';
// models
import Message from '../models/Message';
import Channel from '../models/Channel';
// virtual windowing tool
import { Virtuoso, VirtuosoMethods } from 'react-virtuoso';
// Redux
import { connect } from 'react-redux';
import { Typography } from '@material-ui/core';
// socket.io
import io from 'socket.io-client';

const drawerWidth = 240;

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      display: 'flex',
      height: '100%',
      width: '100vw',
    },
    drawer: {
      [theme.breakpoints.up('sm')]: {
        width: drawerWidth,
        flexShrink: 0,
      },
      zIndex: 1,
    },
    appBar: {
      [theme.breakpoints.up('sm')]: {
        width: `calc(100% - ${drawerWidth}px)`,
        marginLeft: drawerWidth,
      },
    },
    menuButton: {
      marginRight: theme.spacing(2),
      [theme.breakpoints.up('sm')]: {
        display: 'none',
      },
    },
    channelHeader: {
      height: '5vh',
      padding: theme.spacing(2),
    },
    // necessary for content to be below app bar
    toolbar: theme.mixins.toolbar,
    drawerPaper: {
      width: drawerWidth,
      position: 'relative',
    },
    content: {
      flexGrow: 1,
      height: '85vh',
    },
    messageSendWindow: {
      left: drawerWidth,
    },
    image: {
      image: `url(${state})`,
      backgroundRepeat: 'no-repeat',
      backgroundSize: 'auto',
    },
  })
);

const MessageCenter = (props: any) => {
  const classes = useStyles();
  const [channel, setChannel] = useState((null as unknown) as Channel);
  const [channelList, setChannelList] = useState([] as Channel[]);
  const [messageList, setMessageList] = useState([] as Message[]);

  const [messageReceived, setMessageReceived] = useState(false);
  const [socket, setSocket] = useState(
    (null as unknown) as SocketIOClient.Socket
  );
  const [invisible, setInvisible] = useState([] as boolean[]);
  const virtuoso = useRef<VirtuosoMethods>(null);
  // this will need to be changed a lot so abstract it out some
  const messageView = (
    <Virtuoso
      ref={virtuoso}
      totalCount={messageList.length}
      followOutput={true}
      item={(index) => {
        return messageRow(messageList, index);
      }}
    />
  );
  // make sure the socket is defined
  if (!socket) {
    const newSocket: SocketIOClient.Socket = io('http://localhost:4000/');
    setSocket(newSocket);
  }
  const socketCallback = (message: any) => {
    // this has the IDs instead of objects for user and channel
    if (channel && message.channel === channel._id) {
      // add the message to the list and update
      setMessageList((m) => {
        m.push(message);
        return [...m];
      });
    } else if (channel && message.channel !== channel._id) {
      //determine which channel has been updated and set its invisibility amrker
      let channelWithNotification: number = channelList.findIndex(
        (channelSearch) => {
          return message.channel === channelSearch._id;
        }
      );
      if (channelWithNotification !== -1) {
        handleBadgeVisibility(false, channelWithNotification);
        setMessageReceived(!messageReceived);
      }
    }
  };

  // helper function for setting boolean notifications
  const handleBadgeVisibility = (value: boolean, index: number) => {
    setInvisible((invisible) => {
      invisible[index] = value;
      return invisible;
    });
  };

  // get the list of channels in this organization from the backend and set the default channel to General
  useEffect(() => {
    getChannels(props.auth).then((allChannels) => {
      if (allChannels) {
        setChannelList(allChannels);
        // initialize the array with the amount of channels there are
        const invisibilityArray = Array.apply(
          null,
          Array(allChannels.length)
        ).map(() => true);
        // set it to the state variable
        setInvisible(invisibilityArray);
        // join all of the channels with our socket
        if (socket) {
          for (let i = 0; i < allChannels.length; i++) {
            socket.emit('join', allChannels[i]._id);
          }
        }
        // start off on General
        const outChannel = allChannels.find((val) => {
          if (val.name === 'General') {
            return val;
          }
          return undefined;
        });
        if (outChannel) {
          props.changeChannel(outChannel);
          setChannel(outChannel);
        }
      }
    });
  }, []);

  // set up socket event for getting messages
  useEffect(() => {
    try {
      if (socket) {
        socket.off('message');
        socket.on('message', socketCallback);
      }
    } catch (err) {
      console.log(err);
    }
  }, [messageList]);

  // rerender if we receive a message so the notification pops up
  useEffect(() => {}, [messageReceived]);
  const moveToBottom = () => {
    setTimeout(() => {
      if (virtuoso && virtuoso.current) {
        virtuoso.current.scrollToIndex({
          index: messageList.length + 1,
          align: 'end',
        });
      }
    }, 150);
  };
  // get all of the messages in the current channel
  useEffect(() => {
    if (channel) {
      getMessagesFromChannel(channel, props.auth).then((response) => {
        if (response) {
          const receivedMessagesList = response.data;
          setMessageList(receivedMessagesList);
        }
      });
    }
  }, [channel]);

  const channelDrawer = (
    <>
      <nav className={classes.drawer} aria-label='mailbox folders'>
        <Hidden xsDown implementation='css'>
          <Drawer
            classes={{
              paper: classes.drawerPaper,
            }}
            variant='permanent'
            open>
            <Typography variant='h6' align='center'>
              Channels
            </Typography>
            <Divider />
            <List>
              {channelList.map((item, index) => (
                <ListItem
                  button
                  key={item.name}
                  onClick={() => {
                    props.changeChannel(item);
                    setChannel(item);
                    handleBadgeVisibility(true, index);
                  }}>
                  <ListItemIcon>
                    <StopRoundedIcon />
                  </ListItemIcon>
                  <Badge
                    color='secondary'
                    variant='dot'
                    invisible={invisible[index]}>
                    <ListItemText primary={item.name} />
                  </Badge>
                </ListItem>
              ))}
            </List>
            <Divider />
          </Drawer>
        </Hidden>
      </nav>
    </>
  );

  // goes to the most recent message on page load
  useEffect(moveToBottom, [messageList]);
  // Redirect if not logged in
  if (!props.isAuthenticated) {
    return <Redirect to='/' />;
  }

  return (
    <>
      <NavBar />
      <div className={classes.root}>
        {channelDrawer}
        <main className={classes.content}>
          {/* <div className={classes.root} /> */}
          <div className={classes.channelHeader}>
            {channel && <Typography variant='h4'>{channel.name}</Typography>}
          </div>
          <div
            style={{
              height: '78%',
            }}>
            {messageView}
          </div>
          <SendMessageArea />
        </main>
      </div>
    </>
  );
};
const mapStateToProps = (state: any) => ({
  isAuthenticated: state.auth.isAuthenticated,
  auth: state.auth.token,
  userInfo: state.auth.user,
});

export default connect(mapStateToProps, { changeChannel })(MessageCenter);
