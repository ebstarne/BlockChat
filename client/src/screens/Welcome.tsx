import React, { useEffect } from 'react';
import nc from '../assets/nc.png';
import { makeStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Grid from '@material-ui/core/Grid';
import NavBar from '../components/NavBar';
import { Typography } from '@material-ui/core';
import { Link } from 'react-router-dom';
// Redux
import { getTheme } from '../utils/theme';
import { chooseTheme, resetTheme } from '../redux/ducks/actions/theme';
import { connect } from 'react-redux';

const useStyles = makeStyles((theme) => ({
  root: {
    height: '90vh',
  },
  paper: {
    margin: theme.spacing(8, 4),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'left',
    position: 'absolute',
    right: '1%',
    top: '25%',
  },
  image: {
    backgroundImage: `url(${nc})`,
    backgroundRepeat: 'no-repeat',
    backgroundSize: 'cover',
    backgroundPosition: 'center',
  },
  formControl: {
    minWidth: 195,
  },
}));

const Welcome = (props: any) => {
  const classes = useStyles();
  // reset the theme on load
  useEffect(() => {
    props.resetTheme();
  }, []);
  // get the new theme that we want
  const assignTheme = () => {
    getTheme().then((result) => {
      props.chooseTheme(result);
    });
  };

  return (
    <>
      <NavBar />
      <Grid container className={classes.root}>
        <Grid
          item
          className={classes.image}
          xs={false}
          sm={7}
          md={8}
          lg={9}></Grid>

        <Grid
          className={classes.paper}
          container
          item
          direction='column'
          justify='space-evenly'
          alignItems='flex-start'
          alignContent='flex-end'
          spacing={1}
          xs={12}
          sm={5}
          md={4}
          lg={3}>
          <Grid item>
            <Typography variant='h4'>Welcome</Typography>
          </Grid>
          <Grid item>
            <Link to='/login'>
              <Button
                onClick={assignTheme}
                variant='contained'
                color='primary'
                disableElevation>
                Enter
              </Button>
            </Link>
          </Grid>
        </Grid>
      </Grid>
    </>
  );
};

export default connect(null, { chooseTheme, resetTheme })(Welcome);
