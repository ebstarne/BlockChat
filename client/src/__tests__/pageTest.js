import React from 'react';
import ReactDOM from 'react-dom';
import { act } from 'react-dom/test-utils';

import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { BrowserRouter } from 'react-router-dom';
import { SnackbarProvider } from 'notistack';

import Login from '../screens/Login';
import Welcome from '../screens/Welcome';
import MessageCenter from '../screens/MessageCenter';
import InfoButton from '../components/InfoButton';
import messageRow from '../components/messageRow';
import NavBar from '../components/NavBar';
import RegisterUser from '../components/RegisterUser';
import SendMessageArea from '../components/SendMessageArea';

describe('Login', () => {
  let store;
  let component;

  const mockStore = configureStore([]);

  beforeEach(() => {
    store = mockStore({
      auth: {
        isAuthenticated: 'true',
      },
      theme: {
        org: 'Galactic Empire',
        theme: {
          common: {
            black: '#000',
            white: '#fff',
          },
          background: {
            paper: '#fff',
            default: '#fafafa',
          },
          primary: {
            light: 'rgba(94, 194, 250, 1)',
            main: 'rgba(14, 146, 199, 1)',
            dark: 'rgba(0, 100, 150, 1)',
            contrastText: 'rgba(0, 0, 0, 1)',
          },
          secondary: {
            light: 'rgba(255, 255, 255, 1)',
            main: 'rgba(255, 255, 255, 1)',
            dark: 'rgba(255, 255, 255, 1)',
            contrastText: 'rgba(255, 255, 255, 1)',
          },
          error: {
            light: '#e57373',
            main: '#f44336',
            dark: '#d32f2f',
            contrastText: 'rgba(255, 255, 255, 1)',
          },
          text: {
            primary: 'rgba(0, 0, 0, 0.87)',
            secondary: 'rgba(0, 0, 0, 0.54)',
            disabled: 'rgba(0, 0, 0, 0.38)',
            hint: 'rgba(0, 0, 0, 0.38)',
          },
        },
      },
    });
  });

  it('Test the login page is loaded', () => {
    component = renderer.create(
      <Provider store={store}>
        <SnackbarProvider maxSnack={3}>
          <BrowserRouter>
            <Login />
          </BrowserRouter>
        </SnackbarProvider>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  //   it("Test the actual page is loaded", () => {
  //     component = renderer.create(
  //       <Provider store={store}>
  //         <SnackbarProvider maxSnack={3}>
  //           <BrowserRouter>
  //             <MessageCenter />
  //           </BrowserRouter>
  //         </SnackbarProvider>
  //       </Provider>
  //     );
  //     expect(component.toJSON()).toMatchSnapshot();
  //   });

  it('Test the welcome page is loaded', () => {
    component = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Welcome />
        </BrowserRouter>
      </Provider>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});
