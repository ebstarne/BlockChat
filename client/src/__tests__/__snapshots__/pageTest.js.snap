// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Test the login page is loaded 1`] = `null`;

exports[`Login Test the welcome page is loaded 1`] = `
Array [
  <header
    className="MuiPaper-root MuiAppBar-root MuiAppBar-positionRelative MuiAppBar-colorPrimary MuiPaper-elevation4"
    style={
      Object {
        "boxShadow": "none",
      }
    }
  >
    <div
      className="MuiToolbar-root MuiToolbar-regular MuiToolbar-gutters"
    >
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-space-between"
      >
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <img
            alt="State"
            height="64"
            src="state.png"
            style={
              Object {
                "alignContent": "left",
              }
            }
            width="143"
          />
        </div>
        <div
          className="MuiGrid-root MuiGrid-item"
        >
          <div>
            <button
              aria-haspopup="true"
              aria-label="additional info for screen"
              className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit"
              disabled={false}
              onBlur={[Function]}
              onClick={[Function]}
              onDragLeave={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex={0}
              type="button"
            >
              <span
                className="MuiIconButton-label"
              >
                <svg
                  aria-hidden={true}
                  className="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                  />
                </svg>
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </header>,
  <div
    className="MuiGrid-root makeStyles-root-5 MuiGrid-container"
  >
    <div
      className="MuiGrid-root makeStyles-image-7 MuiGrid-item MuiGrid-grid-sm-7 MuiGrid-grid-md-8 MuiGrid-grid-lg-9"
    />
    <div
      className="MuiGrid-root makeStyles-paper-6 MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-item MuiGrid-direction-xs-column MuiGrid-align-items-xs-flex-start MuiGrid-align-content-xs-flex-end MuiGrid-justify-xs-space-evenly MuiGrid-grid-xs-12 MuiGrid-grid-sm-5 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <h4
          className="MuiTypography-root MuiTypography-h4"
        >
          Welcome
        </h4>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item"
      >
        <a
          href="/login"
          onClick={[Function]}
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-disableElevation"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Enter
            </span>
          </button>
        </a>
      </div>
    </div>
  </div>,
]
`;
